python/cudf_polars/cudf_polars/experimental/scheduler.py

```diff
@@ -0,0 +1,148 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Synchronous task scheduler."""
+
+from __future__ import annotations
+
+from collections.abc import MutableMapping
+from typing import TYPE_CHECKING, Any, TypeVar
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+    from typing import TypeAlias
+
+
+Key: TypeAlias = str | tuple[str, int]
+Graph: TypeAlias = MutableMapping[Key, Any]
+T_ = TypeVar("T_")
+
+
+# NOTE: This is a slimmed-down version of the single-threaded
+# (synchronous) scheduler in `dask.core`.
+#
+# Key Differences:
+# * We do not allow a task to contain a list of key names.
+#   Keys must be distinct elements of the task.
+
+
+def istask(x: Any) -> bool:
+    """Check if x is a callable task."""
+    return bool(type(x) is tuple and x and callable(x[0]))
+
+
+def _execute_task(arg: Any, cache: MutableMapping) -> Any:
+    """Execute a compute task."""
+    if istask(arg):
+        return arg[0](*(_execute_task(a, cache) for a in arg[1:]))
+    else:
+        try:
+            return cache.get(arg, arg)
+        except TypeError:  # Unhashable task argument
+            return arg
+
+
+def required_keys(key: Key, graph: Graph) -> list[Key]:
+    """
+    Return the dependencies to extract a key from the graph.
+
+    Parameters
+    ----------
+    key
+        Root key we want to extract.
+    graph
+        The full task graph.
+
+    Returns
+    -------
+    List of other keys needed to extract ``key``.
+    """
+    keys = []
+    tasks = [graph[key]]
+    while tasks:
+        work = []
+        for w in tasks:
+            if istask(w):
+                work.extend(w[1:])
+            else:
+                try:
+                    if w in graph:
+                        keys.append(w)
+                except TypeError:  # not hashable
+                    pass
+        tasks = work
+    return keys
+
+
+def toposort(graph: Graph) -> list[Key]:
+    """Return a list of task keys sorted in topological order."""
+    # Stack-based depth-first search traversal. This is based on Tarjan's
+    # method for topological sorting (see wikipedia for pseudocode)
+    ordered: list[Key] = []
+    completed: set[Key] = set()
+    seen: set[Key] = set()
+
+    dependencies: Mapping[Key, list[Key]] = {k: required_keys(k, graph) for k in graph}
```

Comment:
This is accidentally quadratic in the size of the graph: to obtain the `required_keys` of the final result, we must traverse the entire input. To obtain the required keys of the direct dependencies, we must traverse almost the entire input, etc...

Reply:
Yes, your understanding is correct. Note that we are now only calling this function once for each key in the graph (to build a global dependency graph, and ref-counting dictionary).